aviary/subsystems/propulsion/hamilton_standard.py

```diff
@@ -0,0 +1,932 @@
+        partials["tip_mach",
+                 Dynamic.Mission.PROPELLER_TIP_SPEED] = sqrt_temp_ratio / 1118.21948771
+        partials["tip_mach", Dynamic.Mission.TEMPERATURE] = - \
+            tipspd * sqrt_temp_ratio/(1118.21948771*2*temp)
+        partials["adv_ratio", Dynamic.Mission.VELOCITY] = 5.309 / tipspd
+        partials["adv_ratio", Dynamic.Mission.PROPELLER_TIP_SPEED] = - \
+            5.309 * vktas / (tipspd * tipspd)
+        partials["power_coefficient", Dynamic.Mission.SHAFT_POWER] = 10.E10 * \
+            RHO_SEA_LEVEL_ENGLISH / (rho * 2.*tipspd**3*diam_prop**2*6966.)
+        partials["power_coefficient", Dynamic.Mission.DENSITY] = -10.E10 * shp * \
+            RHO_SEA_LEVEL_ENGLISH / (rho * rho * 2.*tipspd**3*diam_prop**2*6966.)
+        partials["power_coefficient", Dynamic.Mission.PROPELLER_TIP_SPEED] = -3 * \
+            10.E10 * shp * RHO_SEA_LEVEL_ENGLISH / \
+            (rho * 2.*tipspd**4*diam_prop**2*6966.)
+        partials["power_coefficient", Aircraft.Engine.PROPELLER_DIAMETER] = -2 * \
+            10.E10 * shp * RHO_SEA_LEVEL_ENGLISH / \
+            (rho * 2.*tipspd**3*diam_prop**3*6966.)
+
+
+# perfm will probably be a group
+class HamiltonStandard(om.ExplicitComponent):
+    """
+    This is Hamilton Standard component rewritten from Fortran code. 
+    The original documentation is available at 
+    https://ntrs.nasa.gov/api/citations/19720010354/downloads/19720010354.pdf
+    It computes the thrust coefficient of a propeller.
+    """
+
+    def initialize(self):
+        self.options.declare('num_blades', default=2,
+                             desc='number of blades per propeller')
+        self.options.declare(Settings.VERBOSITY,
+                             default=Verbosity.QUIET, desc='print control')
+        self.options.declare(
+            'aviary_options', types=AviaryValues,
+            desc='collection of Aircraft/Mission specific options')
+        self.options.declare('num_nodes', default=1, types=int)
+
+    def setup(self):
+        nn = self.options['num_nodes']
+
+        self.add_input('power_coefficient', val=np.zeros(nn), units='unitless')
+        self.add_input('adv_ratio', val=np.zeros(nn), units='unitless')
+        self.add_input(Dynamic.Mission.MACH, val=np.zeros(nn), units='unitless')
+        self.add_input('tip_mach', val=np.zeros(nn), units='unitless')
+        self.add_input(Aircraft.Engine.PROPELLER_ACTIVITY_FACTOR, val=0.0,
+                       units='unitless')  # Actitivty Factor per Blade
+        self.add_input(Aircraft.Engine.PROPELLER_INTEGRATED_LIFT_COEFFICENT,
+                       val=0.0, units='unitless')  # blade integrated lift coeff
+
+        self.add_output('thrust_coefficient', val=np.zeros(nn), units='unitless')
+        self.add_output('ang_blade', val=np.zeros(nn), units='deg')
+        self.add_output('comp_tip_loss_factor', val=np.zeros(nn), units='unitless')
+
+    def compute(self, inputs, outputs):
```

Comment:
A lot of interesting things here. Since this is in the mission, it will need derivatives. It is a very complicated algorithm with a couple of different interp steps.  First pass at derivs would be "fd", of course. There are some other options.

1. Use complex step --> To do this, will need to make this code complex-safe.  Just at a quick glance, i can see that all the `np.zeros(7)` need to have a dtype added, using the dtype from any input. That way, they will allocate complex if the inputs are complex. I am sure there are a few other things to change.
2. Experiment with jax.  This introduces a dependency, unless it is optional, and falls back on cs if it can't import jax.
3. Turn the algorithm into mostly vectorized numpy calls, and then analytically differentiate, which should be "easier" than differentiating the looped algorithm.   This route is definitely the hardest of the bunch.

I can help on this, at the very least evaluating the feasibility of 3.

No action needed for the PR, except probably to declare the FD partials.

Reply:
All of this makes sense, thank you Ken! Xun and I just talked through options 1-3 and we understand them well. As part of this PR, I will add just the FD partials declaration.